netlify/next-runtime
# feat: add edge middleware support to `ntl dev`

### Summary
Adds support for running middleware using edge functions with `ntl dev`. 

When `next dev` is run, it doesn't emit any of the compiled files that we need to run middleware via our normal edge function. This means that currently we need to run `ntl build` before `ntl dev`, and changes would require a recompiled. This is not good DX: we want to be able to run `ntl dev` without building or linking the site, and have it update middleware when it is changed.

In this PR this is implemented with a special dev-only edge function which runs the middleware directly, rather than via the Next server. While importing the middleware.ts file directly works in simple cases, it breaks if the middleware imports anything else, because the imports won't be Deno-compatible. We handle this by running `esbuild --watch` on the middleware, and importing the compiled middleware. It still works if the middleware hasn't been created yet because esbuild will compile it when it is.

The wrapper edge function checks at request time whether the compiled middleware exists. If not, it just silently returns as there is no middleware. If it does exist then it dynamically imports it, then executes it. This means that it will always have the latest version without needing to reload the edge function.

The edge function is generated in a new `onDev` hook, which also spawns `esbuild`. A regular build will delete the dev edge function to ensure it's never deployed, and the dev hook deletes the production edge functions to ensure they don't conflict.

This needs e2e tests, which will follow later: #1553

### Test plan

1. Check out repo
2. Run npm i in root
3. Install latest version of CLI once [11.0.0](https://github.com/netlify/cli/pull/4970) is released
4. Edit `demos/middleware/netlify.toml` to switch plugin package names
5. Run `npm i` and `ntl dev` in middleware demo
6. Load site and check the "rewrite static" demo. You should see a cat and your city.

### Relevant links (GitHub issues, Notion docs, etc.) or a picture of cute animal
![capy lifeguard](https://user-images.githubusercontent.com/213306/185384234-0d515904-e40a-48b0-9631-9c003fe9faf2.png)

### Standard checks:




---

🧪 Once merged, make sure to update the version if needed and that it was published correctly.
